src/main/java/net/dries007/tfc/common/blocks/plant/KelpTreeFlowerBlock.java

```diff
@@ -0,0 +1,355 @@
+            {
+                return false;
+            }
+        }
+        return true;
+    }
+
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public BlockState updateShape(BlockState stateIn, Direction facing, BlockState facingState, IWorld worldIn, BlockPos currentPos, BlockPos facingPos)
+    {
+        final Fluid containedFluid = stateIn.getValue(getFluidProperty()).getFluid();
+        if (containedFluid != Fluids.EMPTY)
+        {
+            worldIn.getLiquidTicks().scheduleTick(currentPos, containedFluid, containedFluid.getTickDelay(worldIn));
+        }
+        if (facing != Direction.UP && !stateIn.canSurvive(worldIn, currentPos))
+        {
+            worldIn.getBlockTicks().scheduleTick(currentPos, this, 1);
+            return Blocks.AIR.defaultBlockState();
+        }
+        return stateIn;
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public boolean canSurvive(BlockState state, IWorldReader worldIn, BlockPos pos)
+    {
+        KelpTreeBlock body = (KelpTreeBlock) getBodyBlockAbstract().get();
+
+        BlockState blockstate = worldIn.getBlockState(pos.below());
+        if (blockstate.getBlock() != body && !blockstate.is(TFCTags.Blocks.SEA_BUSH_PLANTABLE_ON))
+        {
+            if (!isEmptyWaterBlock(worldIn, pos.below()))
+            {
+                return false;
+            }
+            else
+            {
+                boolean isValid = false;
+                for (Direction direction : Direction.Plane.HORIZONTAL)
+                {
+                    BlockState relativeState = worldIn.getBlockState(pos.relative(direction));
+                    if (relativeState.is(body))
+                    {
+                        if (isValid)
+                        {
+                            return false;
+                        }
+
+                        isValid = true;
+                    }
+                    else if (!isEmptyWaterBlock(worldIn, pos.relative(direction)))
+                    {
+                        return false;
+                    }
+                }
+
+                return isValid;
+            }
+        }
+        else
+        {
+            return true;
+        }
+    }
+
+    @Override
+    protected void createBlockStateDefinition(StateContainer.Builder<Block, BlockState> builder)
+    {
+        builder.add(getFluidProperty(), AGE);
+    }
+
+    public void generatePlant(IWorld worldIn, BlockPos pos, Random rand, int maxHorizontalDistance, Fluid fluid)
+    {
+        if (!getFluidProperty().canContain(fluid))
+            return;
+        setBodyBlockWithFluid(worldIn, pos, fluid);
+        growTreeRecursive(worldIn, pos, rand, pos, maxHorizontalDistance, 0, fluid);
+    }
+
+    public void growTreeRecursive(IWorld worldIn, BlockPos branchPos, Random rand, BlockPos originalBranchPos, int maxHorizontalDistance, int iterations, Fluid fluid)
```

Comment:
Just taking a guess this looks like mojang copy paste - if so, always good to provide a javadoc with an `@link` to the copy pasta'd code so we can check if/when it changes across versions, and cross reference what it's supposed to be.